ocaml/xapi/stream_vdi.ml

```diff
@@ -194,95 +195,174 @@ let get_chunk_numbers_in_increasing_order descriptor_list offset =
   let chunks = process [] offset descriptor_list in
   List.rev chunks
 
+let get_allocated_chunks_from_clusters cluster_size cluster_list =
+  let chunk_size = Int64.to_int chunk_size in
+  let chunks_in_cluster = (cluster_size + chunk_size - 1) / chunk_size in
+  let set =
+    List.fold_left
+      (fun set cluster_no ->
+        let cluster_offset = cluster_no * cluster_size in
+        let chunk_no = cluster_offset / chunk_size in
+        let chunks_to_add =
+          Seq.init chunks_in_cluster (fun i -> chunk_no + i)
+        in
+        ChunkSet.add_seq chunks_to_add set
+      )
+      ChunkSet.empty cluster_list
+  in
+  set
+
 let send_one ofd (__context : Context.t) rpc session_id progress refresh_session
     (prefix, vdi_ref, _size) =
   let size = Db.VDI.get_virtual_size ~__context ~self:vdi_ref in
+  (* Remember when we last wrote something so that we can work around firewalls which close 'idle' connections *)
+  let last_transmission_time = ref 0L in
```

Comment:
Mtime_clock.counter could be used instead, I think it would improve readability:

```ocaml
let last_transmission_time = ref (Mtime_clock.counter ())

...

    if write_check buffer first_or_last then (
      last_transmission_time := Mtime_clock.counter () ;

...

  let need_to_retransmit time_since = Mtime.Span.(is_longer ~than:(5 * s) time_since) in

...

              if
               need_to_retransmit (Mtime_clock.count time_since_transmission)
                  ...
              then (

...
```